torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1786,13 +2023,36 @@ def _update_p_data(self, p, output_tensor: torch.Tensor) -> None:
         p.data = p.data[: p._orig_size.numel()].view(p._orig_size)  # type: ignore[attr-defined]
 
     @torch.no_grad()
-    def _rebuild_full_params(self) -> None:
+    def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
         """
         Gather all shards of params.
         """
+        # summon_full_params must do a full precision rebuild even under mixed
+        # precision, because it is used for e.g. checkpoint where we'd like to
+        # checkpoint in full precision.
+        force_full_precision = (self.training_state == TrainingState_.SUMMON_FULL_PARAMS)
+        # full param output tensors and a flag indicating whether
+        # summon_full_params can free them or not. It is possible that we can't
+        # free the full param, which currently occurs when the returned
+        # parameter points to the unsharded param when world_size == 1, or when
+        # we're returning the full parameter and reshard_after_forward=False
+        # (because we need to ensure p._full_param_padded stays intact)
+        output_tensors: List[Tuple[torch.Tensor, bool]] = []
         with torch.cuda.stream(self._streams["all_gather"]):
             for p in self.params:
-                if self.cpu_offload.offload_params:
+                mixed_precision_cast_ran = (
+                    self.mixed_precision is not None and not force_full_precision
+                )
+                if mixed_precision_cast_ran:
```

Comment:
nit: Coalesce two `if` blocks?

Reply:
Sorry, the next if is just debug vestige, it should not be needed